### PR TITLE
Fix sequence name 

### LINKED ISSUE #81: NameError: instance variable @sequence_name not defined
After upgrading to ros-apartment from apartment, I get the above error at times:

```
[GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:124 :in `remove_instance_variable`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:124 :in `block in reset_sequence_names`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:123 :in `each`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:123 :in `reset_sequence_names`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:40 :in `reset`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:30 :in `initialize`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:11 :in `new`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:11 :in `postgresql_adapter`
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/tenant.rb:56 :in `adapter`
/home/app_au/.rbenv/versions/2.6.5/lib/ruby/2.6.0/forwardable.rb:224 :in `switch!`
 [PROJECT_ROOT]/lib/country_manager.rb:188 :in `country=`
 [PROJECT_ROOT]/config/initializers/sidekiq.rb:55 :in `set_country_from_job`
 [PROJECT_ROOT]/config/initializers/sidekiq.rb:27 :in `call`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/middleware/chain.rb:140 :in `block in invoke`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/middleware/chain.rb:140 :in `block in invoke`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/middleware/chain.rb:143 :in `invoke`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:163 :in `block in process`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:136 :in `block (6 levels) in dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/job_retry.rb:111 :in `local`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:135 :in `block (5 levels) in dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/rails.rb:43 :in `block in call`
 [GEM_ROOT]/gems/activesupport-5.2.4.2/lib/active_support/execution_wrapper.rb:87 :in `wrap`
 [GEM_ROOT]/gems/activesupport-5.2.4.2/lib/active_support/reloader.rb:73 :in `block in wrap`
 [GEM_ROOT]/gems/activesupport-5.2.4.2/lib/active_support/execution_wrapper.rb:87 :in `wrap`
 [GEM_ROOT]/gems/activesupport-5.2.4.2/lib/active_support/reloader.rb:72 :in `wrap`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/rails.rb:42 :in `call`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:131 :in `block (4 levels) in dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:257 :in `stats`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:126 :in `block (3 levels) in dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/job_logger.rb:13 :in `call`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:125 :in `block (2 levels) in dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/job_retry.rb:78 :in `global`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:124 :in `block in dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/logger.rb:10 :in `with`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/job_logger.rb:33 :in `prepare`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:123 :in `dispatch`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:162 :in `process`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:78 :in `process_one`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/processor.rb:68 :in `run`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/util.rb:15 :in `watchdog`
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/util.rb:24 :in `block in safe_thread`More (36 lines)
Nested Exceptions
NameError: instance variable @sequence_name not defined
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:124 :in `remove_instance_variable`
 Sidekiq::JobRetry::Handled: Sidekiq::JobRetry::Handled
 [GEM_ROOT]/gems/sidekiq-6.0.7/lib/sidekiq/job_retry.rb:99 :in `rescue in global`
 NameError: instance variable @sequence_name not defined
 [GEM_ROOT]/gems/ros-apartment-2.6.1/lib/apartment/adapters/postgresql_adapter.rb:124 :in `remove_instance_variable`
```

Rails 5.2 and latest ros-apartment from rubygems.

Any idea what it could be?

### LINKED ISSUE #161: Postgres sequence_name correction is not thread safe
While reviewing this gem I noticed that https://github.com/rails-on-services/apartment/commit/f8eefc43767fd94a190b182d4ef58178bebebd11 added some changes to the Postgres Adapter. It appears that they are not thread safe.

## Steps to reproduce

Given two tenants - `test1` and `test2`, and a `Jobseeker` model, the following code eventually fails:  

```ruby
def check_sequence_name(tenant)
  Apartment::Tenant.switch(tenant) do
    sequence_name = Jobseeker.sequence_name
    unless sequence_name.to_s.starts_with?(tenant)
      raise "Invalid sequence name for #{tenant}: #{sequence_name}"
    end
  end
end

Thread.new do
  while true do
    check_sequence_name('test1')
  end
end

Thread.new do
  while true do
    check_sequence_name('test2')
  end
end

#<Thread:0x00005557dd7d22d0@(pry):14 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	4: from (pry):16:in `block in <main>'
	3: from (pry):2:in `check_sequence_name'
	2: from /usr/local/lib/ruby/2.6.0/forwardable.rb:230:in `switch'
	1: from /usr/local/bundle/gems/ros-apartment-2.9.0/lib/apartment/adapters/abstract_adapter.rb:89:in `switch'
(pry):5:in `block in check_sequence_name': Invalid sequence name for test2: test1.jobseekers_id_seq (RuntimeError)
```

## Expected behavior

No exceptions should be raised when running the above test example.

## Actual behavior

The `reset_sequence_names` method modifies sequence names of `ActiveRecord::Base.descendants` without considering thread safety: https://github.com/rails-on-services/apartment/blob/f79ec201e93c2517a2c1dcde92ffa155ebf41dd4/lib/apartment/adapters/postgresql_adapter.rb#L133-L149

## System configuration

* Database: Postgres 12.2

* Apartment version: 2.9.0

* Apartment config (in `config/initializers/apartment.rb` or so):

```ruby
Apartment.configure do |config|
  config.excluded_models = %w{Tenant}
  config.tenant_names = lambda{ Tenant.pluck(:database) }
  config.persistent_schemas = ['shared_extensions']
end
```

* Rails (or ActiveRecord) version: 5.2.4

* Ruby version: 2.6.0

P.S. Thank you for putting so much effort into maintaining this gem.

### PR DESCRIPTION
**Pull request by rpbaltazar**
Resolves #161
Resolves #81 

- Copied @fsateler solution from the [draft](https://github.com/rails-on-services/apartment/pull/143/files) PR to this repo
- Added test coverage to the sceario of tenant switch within a block
- Updated specs to make use of `Apartment::Tenant.adapter` to avoid creating two instances of the adapter that wouldn't be in sync

### CODE REVIEW

**Comment by rpbaltazar** [comment]
I'm trying to address [this concern](https://github.com/rails-on-services/apartment/pull/143#discussion_r552701204)

File: spec/examples/schema_adapter_examples.rb [hunk 1]
```diff
@@ -120,10 +126,12 @@
       subject.switch(schema1) do
         expect(connection.schema_search_path).to start_with %("#{schema1}")
         expect(User.sequence_name).to eq "#{schema1}.#{User.table_name}_id_seq"
```
Comment (rpbaltazar) [inline spec/examples/schema_adapter_examples.rb:128]: Based on the expectations and my understanding of your proposed solution, shouldn't we update this test to ensure that User's sequence name does not have the schema set inside nor outside the block?
@fsateler

File: spec/examples/schema_adapter_examples.rb [hunk 2]
```diff
@@ -120,10 +126,12 @@
...
+        expect(Company.sequence_name).to eq "#{public_schema}.#{Company.table_name}_id_seq"
       end
 
       expect(connection.schema_search_path).to start_with %("#{public_schema}")
-      expect(User.sequence_name).to eq "#{public_schema}.#{User.table_name}_id_seq"
+      expect(User.sequence_name).to eq "#{User.table_name}_id_seq"
```
Comment (rpbaltazar) [inline spec/examples/schema_adapter_examples.rb:133]: And this test currently fails because of my previous note. the sequence name for users is set inside the block and cached, therefore then it's giving the wrong one outside the block

Comment (rpbaltazar) [reply spec/examples/schema_adapter_examples.rb:139]: Ignore my comment. I found that the issue was that we were initializing an adapter instance by calling `Apartment::Tenant.postgresql_adapter config` but not storing it in the thread current variable. and then when accessing Apartment::Tenant methods, a new adapter was being initialized thus the apartment tenant current was different

**Comment by rpbaltazar** [comment]
@fsateler if you have the time, do you mind taking a look at this PR?

File: lib/apartment.rb [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 require_relative 'apartment/log_subscriber'
 
+require_relative 'apartment/active_record/postgresql_adapter'
```
Comment (fsateler) [inline lib/apartment.rb:11]: This would force the loading of the pg adapter. Are other dbs no longer supported? This was my comment on https://github.com/rails-on-services/apartment/pull/143#discussion_r552701204

File: lib/apartment/active_record/postgresql_adapter.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+# NOTE: This patch is meant to remove any schema_prefix appart from the ones for
+# excluded models. The schema_prefix would be resolved by apartment's setting
+# of search path
+module Apartment::PostgreSqlAdapterPatch
+  def default_sequence_name(table, _column)
+    res = super
+    schema_prefix = "#{Apartment::Tenant.current}."
+
+    if res&.starts_with?(schema_prefix)
+      if Apartment.excluded_models.none? { |m| m.constantize.table_name == table }
```
Comment (fsateler) [inline lib/apartment/active_record/postgresql_adapter.rb:14]: I find it hard to parse `if/else` with negatives (`none?`). I would use `any?` and swap the bodies

File: lib/apartment/active_record/postgresql_adapter.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        res.delete_prefix!(schema_prefix)
+      else
+        res.sub!(schema_prefix, "#{Apartment::Tenant.default_tenant}.")
```
Comment (fsateler) [inline lib/apartment/active_record/postgresql_adapter.rb:17]: This causes extra work when `default_tenant == current`

**Review by fsateler — CHANGES_REQUESTED** [review]
I think the only dealbraker is that this would force the loading of pg adatper, which wouldn't exist on non-pg databases.

File: lib/apartment/active_record/postgresql_adapter.rb (hunk 5 above)
Comment (fsateler) [reply lib/apartment/active_record/postgresql_adapter.rb:17]: On second thought, this is not strictly correct, because it should happen even if the sequence name is prefixed with some other scheam. No idea if that can happen though :thinking:

File: lib/apartment.rb (hunk 3 above)
Comment (rpbaltazar) [reply lib/apartment.rb:11]: Yes. I have not looked at how can we fix this and test it on a non postgresql db. I currently run only postgres so it's also hard for me to test these details on all dbs. I've been tempted on starting refactoring the gem to have one gem per adapter and a core, but that's discussion for another day.

File: lib/apartment/active_record/postgresql_adapter.rb (hunk 5 above)
Comment (rpbaltazar) [reply lib/apartment/active_record/postgresql_adapter.rb:17]: I thought of that when i was writing this and then forgot to address it. tks

File: lib/apartment.rb (hunk 3 above)
Comment (fsateler) [reply lib/apartment.rb:11]: perhaps we can go the "easy" route:

```ruby
if defined?(PG)
  require_relative 'apartment/active_record/postgresql_adapter'
end
```

Perhaps in an initializer so that this happens after all gems are loaded?

Comment (rpbaltazar) [reply lib/apartment.rb]: I think i have a clearer working solution. We only require the adapter once we eval the config. I'm requiring this patch after requiring the postgres adapter

File: lib/apartment/active_record/postgresql_adapter.rb (hunk 5 above)
Comment (rpbaltazar) [reply lib/apartment/active_record/postgresql_adapter.rb]: I agree w your second concern. addressed it as well

**Comment by rpbaltazar** [comment]
@marksiemers and @citrus

I have been working on this with @fsateler and just remembered your request to be able to switch to multiple schemas:

- https://github.com/rails-on-services/apartment/pull/112
- https://github.com/rails-on-services/apartment/pull/154

The solution here assumes that the search path will resolve the table's sequence properly and therefore tries to remove the prefix.
I wonder what is the sequence name in your use cases with multiple tenants. Does it return the first schema name that it finds? I think that this path needs to be looked by one of you as I don't have a realistic use case scenario

**Comment by rpbaltazar** [comment]
I'll merge this one first. We can look at your specific use case after

**Comment by marksiemers** [comment]
@rpbaltazar - Thanks for calling this out. I don't work at the company anymore where we used Apartment, so I don't have the system/codebase in place that has the real-world use case. I can offer some info from the postgres docs.

Your question:
> I wonder what is the sequence name in your use cases with multiple tenants. Does it return the first schema name that it finds? I think that this path needs to be looked by one of you as I don't have a realistic use case scenario

From docs:
> The first schema named in the search path is called the current schema. Aside from being the first schema searched, it is also the schema in which new tables will be created
> ,,,
> The first schema in the search path that exists is the default location for creating new objects. That is the reason that by default objects are created in the public schema. When objects are referenced in any other context without schema qualification (table modification, data modification, or query commands) the search path is traversed until a matching object is found.
> ,,,
> The search path works in the same way for data type names, function names, and operator names as it does for table names.

Source: https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PATH

I assume this applies to sequences as well. The first named schema in the search path is where any new object will be created.

@ryanswood - Is this something you can check into or confirm?

**Comment by ryanswood** [comment]
> The solution here assumes that the search path will resolve the table's sequence properly and therefore tries to remove the prefix. I wonder what is the sequence name in your use cases with multiple tenants. Does it return the first schema name that it finds? I think that this path needs to be looked by one of you as I don't have a realistic use case scenario

@rpbaltazar Thank you so much for remembering our request. I worked with @marksiemers and I am at still at the company where the multi-schema switch feature is still needed.

To answer your question, the value of `sequence_name` is dependent on the existence of the the table in the schema and NOT the existence of the sequence. Postgres will scan the search path until it finds the first schema where the table exists and it returns the sequence name if it exists or nil if it does not exist for the found table.

It appears as if the override of `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter#default_sequence_name` needs to account for the [original's implementation when it returns `nil`](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb#L249). As it stands right now, there is a `NoMethodError` when the model is excluded from Apartment and `res.sub!` is called. Would you like me to research this further and submit a PR?

Quick background: We use the multi-schema switch feature for our data pipeline where we create a temporary schema that partially mirrors the original so that we can mass import data and hide it from UI. There are cases where the duplicate table in the temporary schema does not have its own sequence.
